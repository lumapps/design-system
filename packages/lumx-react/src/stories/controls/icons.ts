import {
    mdiAbTesting,
    mdiAbjadArabic,
    mdiAccount,
    mdiAccountBox,
    mdiAlert,
    mdiAlertCircle,
    mdiArrowDown,
    mdiArrowUp,
    mdiAtom,
    mdiBee,
    mdiBell,
    mdiBullhornOutline,
    mdiCheck,
    mdiCheckCircle,
    mdiChevronDown,
    mdiChevronLeft,
    mdiChevronRight,
    mdiChevronUp,
    mdiClose,
    mdiCloseCircle,
    mdiDelete,
    mdiDotsHorizontal,
    mdiDragVertical,
    mdiEarth,
    mdiEmail,
    mdiEye,
    mdiFileEdit,
    mdiFlag,
    mdiFolder,
    mdiFolderGoogleDrive,
    mdiFoodApple,
    mdiGoogleCirclesExtended,
    mdiHeart,
    mdiHome,
    mdiImageBroken,
    mdiInformation,
    mdiLink,
    mdiMagnifyMinusOutline,
    mdiMagnifyPlusOutline,
    mdiMenuDown,
    mdiMessageTextOutline,
    mdiMinus,
    mdiOpenInNew,
    mdiPauseCircleOutline,
    mdiPencil,
    mdiPlay,
    mdiPlayCircleOutline,
    mdiPlus,
    mdiRadioboxBlank,
    mdiRadioboxMarked,
    mdiReply,
    mdiSend,
    mdiStar,
    mdiTextBox,
    mdiTextBoxPlus,
    mdiTram,
    mdiTranslate,
    mdiViewList,
} from '@lumx/icons';

// Small selection of mdi icons to
export const iconArgType = {
    control: { type: 'select' },
    options: {
        mdiAbTesting,
        mdiAbjadArabic,
        mdiAccount,
        mdiAccountBox,
        mdiAlert,
        mdiAlertCircle,
        mdiArrowDown,
        mdiArrowUp,
        mdiAtom,
        mdiBee,
        mdiBell,
        mdiBullhornOutline,
        mdiCheck,
        mdiCheckCircle,
        mdiChevronDown,
        mdiChevronLeft,
        mdiChevronRight,
        mdiChevronUp,
        mdiClose,
        mdiCloseCircle,
        mdiDelete,
        mdiDotsHorizontal,
        mdiDragVertical,
        mdiEarth,
        mdiEmail,
        mdiEye,
        mdiFileEdit,
        mdiFlag,
        mdiFolder,
        mdiFolderGoogleDrive,
        mdiFoodApple,
        mdiGoogleCirclesExtended,
        mdiHeart,
        mdiHome,
        mdiImageBroken,
        mdiInformation,
        mdiLink,
        mdiMagnifyMinusOutline,
        mdiMagnifyPlusOutline,
        mdiMenuDown,
        mdiMessageTextOutline,
        mdiMinus,
        mdiOpenInNew,
        mdiPauseCircleOutline,
        mdiPencil,
        mdiPlay,
        mdiPlayCircleOutline,
        mdiPlus,
        mdiRadioboxBlank,
        mdiRadioboxMarked,
        mdiReply,
        mdiSend,
        mdiStar,
        mdiTextBox,
        mdiTextBoxPlus,
        mdiTram,
        mdiTranslate,
        mdiViewList,
    },
};
