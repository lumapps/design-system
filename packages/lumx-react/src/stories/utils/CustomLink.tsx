/* eslint-disable jsx-a11y/anchor-has-content,jsx-a11y/anchor-is-valid */
import React from 'react';

/**
 * Example custom link that can be used in linkAs props.
 */
export const CustomLink: any = ({ ...props }) => <a href="#" data-custom-link {...props} />;
