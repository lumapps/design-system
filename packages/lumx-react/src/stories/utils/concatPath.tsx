export type PathPart = string | number;
export type Path = PathPart | Array<PathPart>;

/**
 * Concat flatten object path
 *
 * @example concatPath('foo', 'bar') // => 'foo.bar'
 * @example concatPath(['foo', 0]) // => 'foo[0]'
 * @example concatPath('foo', 0, ['bar']) // => 'foo[0].bar'
 */
export const concatPath = (...prefix: Path[]) => {
    const [first, ...rest] = prefix.flat();
    return rest.reduce<string>((acc, part) => {
        if (typeof part === 'number') return `${acc}[${part}]`;
        return `${acc}.${part}`;
    }, String(first));
};
