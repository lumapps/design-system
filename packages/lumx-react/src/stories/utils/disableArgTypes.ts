export function disableArgTypes(argTypes: string[]) {
    return Object.fromEntries(argTypes.map((argType) => [argType, { table: { disable: true } }]));
}
