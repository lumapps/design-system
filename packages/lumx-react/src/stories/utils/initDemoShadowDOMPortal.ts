/** Init a demo shadow DOM to use as portal container */
export const initDemoShadowDOMPortal = () => {
    const div = document.createElement('div');
    document.body.appendChild(div);
    const container = div.attachShadow({ mode: 'open' });
    const style = document.createElement('style');
    style.innerText = `button { color: red; }`;
    container.appendChild(style);
    return { container, teardown: () => div.remove() };
};
