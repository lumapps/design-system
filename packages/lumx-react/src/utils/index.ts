/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
