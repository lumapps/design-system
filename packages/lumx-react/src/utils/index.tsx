export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
