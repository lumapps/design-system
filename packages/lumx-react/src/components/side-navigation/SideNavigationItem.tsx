import React, { Children, forwardRef, ReactNode } from 'react';

import classNames from 'classnames';
import isEmpty from 'lodash/isEmpty';

import { mdiChevronDown, mdiChevronUp } from '@lumx/icons';

import { Emphasis, Icon, Size, IconButton, IconButtonProps } from '@lumx/react';

import {
    Callback,
    Comp,
    GenericProps,
    getRootClassName,
    handleBasicClasses,
    isComponent,
    onEnterPressed,
} from '@lumx/react/utils';
import { renderLink } from '@lumx/react/utils/renderLink';

/**
 * Defines the props of the component.
 */
export interface SideNavigationItemProps extends GenericProps {
    /** SideNavigationItem elements. */
    children?: ReactNode;
    /** Emphasis variant. */
    emphasis?: Emphasis;
    /** Label content. */
    label: string | ReactNode;
    /** Icon (SVG path). */
    icon?: string;
    /** Whether the component is open or not. */
    isOpen?: boolean;
    /** Whether the component is selected or not. */
    isSelected?: boolean;
    /** Custom react component for the link (can be used to inject react router Link). */
    linkAs?: 'a' | any;
    /** Props to pass to the link (minus those already set by the SideNavigationItem props). */
    linkProps?: React.DetailedHTMLProps<React.AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>;
    /** Props to pass to the toggle button (minus those already set by the SideNavigationItem props). */
    toggleButtonProps: Pick<IconButtonProps, 'label'> &
        Omit<IconButtonProps, 'label' | 'onClick' | 'icon' | 'emphasis' | 'color' | 'size'>;
    /** On action button click callback. */
    onActionClick?(evt: React.MouseEvent): void;
    /** On click callback. */
    onClick?(evt: React.MouseEvent): void;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'SideNavigationItem';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<SideNavigationItemProps> = {
    emphasis: Emphasis.high,
};

/**
 * SideNavigationItem component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const SideNavigationItem: Comp<SideNavigationItemProps, HTMLLIElement> = forwardRef((props, ref) => {
    const {
        children,
        className,
        emphasis,
        icon,
        isOpen,
        isSelected,
        label,
        linkAs,
        linkProps,
        onActionClick,
        onClick,
        toggleButtonProps,
        ...forwardedProps
    } = props;

    const content = children && Children.toArray(children).filter(isComponent(SideNavigationItem));
    const hasContent = !isEmpty(content);
    const shouldSplitActions = Boolean(onActionClick);

    return (
        <li
            ref={ref}
            {...forwardedProps}
            className={classNames(
                className,
                handleBasicClasses({
                    emphasis,
                    isOpen,
                    isSelected,
                    prefix: CLASSNAME,
                }),
            )}
        >
            {shouldSplitActions ? (
                <div className={`${CLASSNAME}__wrapper`}>
                    {renderLink(
                        {
                            linkAs,
                            ...linkProps,
                            className: `${CLASSNAME}__link`,
                            onClick,
                            tabIndex: 0,
                        },
                        icon && <Icon className={`${CLASSNAME}__icon`} icon={icon} size={Size.xs} />,
                        <span>{label}</span>,
                    )}

                    <IconButton
                        {...toggleButtonProps}
                        className={`${CLASSNAME}__toggle`}
                        icon={isOpen ? mdiChevronUp : mdiChevronDown}
                        size={Size.m}
                        emphasis={Emphasis.low}
                        onClick={onActionClick}
                    />
                </div>
            ) : (
                renderLink(
                    {
                        linkAs,
                        ...linkProps,
                        className: `${CLASSNAME}__link`,
                        tabIndex: 0,
                        onClick,
                        onKeyDown: onClick ? onEnterPressed(onClick as Callback) : undefined,
                    },
                    icon && <Icon className={`${CLASSNAME}__icon`} icon={icon} size={Size.xs} />,
                    <span>{label}</span>,
                    hasContent && (
                        <Icon
                            className={`${CLASSNAME}__chevron`}
                            icon={isOpen ? mdiChevronUp : mdiChevronDown}
                            size={Size.xs}
                        />
                    ),
                )
            )}

            {hasContent && isOpen && <ul className={`${CLASSNAME}__children`}>{content}</ul>}
        </li>
    );
});
SideNavigationItem.displayName = COMPONENT_NAME;
SideNavigationItem.className = CLASSNAME;
SideNavigationItem.defaultProps = DEFAULT_PROPS;
