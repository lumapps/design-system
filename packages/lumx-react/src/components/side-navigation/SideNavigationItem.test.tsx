import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';

import { commonTestsSuite, Wrapper } from '@lumx/react/testing/utils';
import { getBasicClass } from '@lumx/react/utils';

import { mdiAccount } from '@lumx/icons';

import { SideNavigationItem, SideNavigationItemProps } from './SideNavigationItem';

const DEFAULT_PROPS = SideNavigationItem.defaultProps as any;
const CLASSNAME = SideNavigationItem.className as string;

type SetupProps = Partial<SideNavigationItemProps>;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = ({ ...propsOverride }: SetupProps = {}, shallowRendering = true) => {
    const props: any = {
        toggleButtonProps: { label: 'Toggle' },
        ...propsOverride,
    };
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;
    const wrapper: Wrapper = renderer(<SideNavigationItem {...props} />);

    return {
        chevron: wrapper.find(`.${CLASSNAME}__chevron`),
        children: wrapper.find(`.${CLASSNAME}__children`),
        icon: wrapper.find(`.${CLASSNAME}__icon`),
        label: wrapper.find(`.${CLASSNAME}__link span`),
        link: wrapper.find(`.${CLASSNAME}__link`),
        props,
        root: wrapper.find(`.${CLASSNAME}`),
        wrapper,
    };
};

describe(`<${SideNavigationItem.displayName}>`, () => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        // Here is an example of a basic rendering check, with snapshot.

        it('should render correctly', () => {
            const { root, wrapper } = setup();
            expect(wrapper).toMatchSnapshot();

            expect(root).toExist();
            expect(root).toHaveClassName(CLASSNAME);
        });

        it('should render correctly with splitted actions', () => {
            const { root, wrapper } = setup({ linkProps: { href: 'http://toto.com' }, onClick: () => null });
            expect(wrapper).toMatchSnapshot();

            expect(root).toExist();
            expect(root).toHaveClassName(CLASSNAME);
        });
    });

    // 2. Test defaultProps value and important props custom values.
    describe('Props', () => {
        // Here are some examples of basic props check.

        it('should use default props', () => {
            const { root } = setup();

            for (const prop of Object.keys(DEFAULT_PROPS)) {
                const className = getBasicClass({ prefix: CLASSNAME, type: prop, value: DEFAULT_PROPS[prop] });
                if (className) {
                    expect(root).toHaveClassName(className);
                }
            }
        });
    });

    // 3. Test events.
    describe('Events', () => {
        const onClick: jest.Mock = jest.fn();

        beforeEach(onClick.mockClear);

        it('should trigger `onClick`', () => {
            const { link } = setup({ onClick }, false);
            link.simulate('click');
            expect(onClick).toHaveBeenCalled();
        });
    });

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', () => {
        // Here is an example of children types check.

        const items = [
            <SideNavigationItem key="a" label="a" toggleButtonProps={{ label: 'Toggle' }} />,
            <SideNavigationItem key="b" label="b" toggleButtonProps={{ label: 'Toggle' }} />,
        ];

        it('should hide chevron when no children are passed', () => {
            const { chevron } = setup({
                children: [],
            });

            expect(chevron).not.toExist();
        });

        it('should show chevron when children are passed', () => {
            const { chevron } = setup({
                children: items,
            });

            expect(chevron).toExist();
        });

        it('should hide children when children are passed and isOpen is false', () => {
            const { children } = setup({
                children: items,
                isOpen: false,
            });

            expect(children).not.toExist();
        });

        it('should show children when children are passed and isOpen is true', () => {
            const { children } = setup({
                children: items,
                isOpen: true,
            });

            expect(children).toExist();
        });

        it('should show icon when provided', () => {
            const { icon } = setup({
                icon: mdiAccount,
            });

            expect(icon).toExist();
        });

        it('should add props to link when provided', () => {
            const { link } = setup({
                linkProps: {
                    href: '/',
                },
            });

            expect(link.prop('href')).toEqual('/');
        });
    });

    // 5. Test state => no state

    // Common tests suite.
    commonTestsSuite(setup, { className: 'root', prop: 'root' }, { className: CLASSNAME });
});
