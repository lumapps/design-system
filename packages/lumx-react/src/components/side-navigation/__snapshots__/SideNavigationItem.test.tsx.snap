// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SideNavigationItem> Snapshots and structure should render correctly 1`] = `
<li
  className="lumx-side-navigation-item lumx-side-navigation-item--emphasis-high"
>
  <a
    className="lumx-side-navigation-item__link"
    tabIndex={0}
  >
    <span />
  </a>
</li>
`;

exports[`<SideNavigationItem> Snapshots and structure should render correctly with splitted actions 1`] = `
<li
  className="lumx-side-navigation-item lumx-side-navigation-item--emphasis-high"
>
  <a
    className="lumx-side-navigation-item__link"
    href="http://toto.com"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    <span />
  </a>
</li>
`;
